llrt_modules/src/modules/zlib.rs

```diff
@@ -0,0 +1,293 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+use std::io::Read;
+
+use brotlic::{CompressorReader as BrotliEncoder, DecompressorReader as BrotliDecoder};
+use flate2::{
+    read::{DeflateDecoder, DeflateEncoder, GzDecoder, GzEncoder, ZlibDecoder, ZlibEncoder},
+    Compression,
+};
+use llrt_utils::{ctx::CtxExtension, module::export_default, object::ObjectExt, result::ResultExt};
+use rquickjs::function::Func;
+use rquickjs::{
+    module::{Declarations, Exports, ModuleDef},
+    prelude::{Opt, Rest},
+    Ctx, Error, Exception, FromJs, Function, IntoJs, Null, Object, Result, Value,
+};
+use zstd::{
+    stream::read::{Decoder as ZstdDecoder, Encoder as ZstdEncoder},
+    zstd_safe::CompressionLevel,
+};
+
+use crate::{utils::array_buffer::ArrayBufferView, ModuleInfo};
+
+use super::buffer::Buffer;
+
+macro_rules! define_sync_function {
+    ($fn_name:ident, $converter:expr, $command:expr) => {
+        pub fn $fn_name<'js>(
+            ctx: Ctx<'js>,
+            value: Value<'js>,
+            options: Opt<Object<'js>>,
+        ) -> Result<Value<'js>> {
+            $converter(ctx.clone(), value, options, $command)
+        }
+    };
+}
+
+macro_rules! define_cb_function {
+    ($fn_name:ident, $converter:expr, $command:expr) => {
+        pub fn $fn_name<'js>(
+            ctx: Ctx<'js>,
+            value: Value<'js>,
+            args: Rest<Object<'js>>,
+        ) -> Result<()> {
+            let mut args_iter = args.0.into_iter().rev();
+            let cb: Function = args_iter
+                .next()
+                .and_then(|v| <rquickjs::Value<'_> as Clone>::clone(&v).into_function())
+                .or_throw_msg(&ctx, "Callback required")?;
```

Comment:
Doesn't `into_function() `automatically clone? Also error message is a bit misleading. "Callback parameter is not a function" or similar is more clear.

Reply:
Sorry, the type of the `options` parameter was mixed between `Object<'js>` and `Value<'js>` and needed to be converted. We have simplified the process by unifying it with `Values<'js>`.

I also reviewed the error messages.